rfcs/20221124-monorepo.md

```diff
@@ -0,0 +1,179 @@
+↖️ Table of Contents
+
+# RFC: Monorepo
+
+**Status:** 🚧 WIP, comments are welcome nonetheless
+
+## Reviewers
+
+- [ ] @zackkrida
+- [ ] @sarayourfriend
+
+## Rationale
+
+For a comprehensive discussion about the pros, the cons and the counterpoints to each see [discussion](https://github.com/WordPress/openverse/issues/192). This is not the purpose of this RFC.
+
+This RFC summarily lists the benefits and then, with the twin assumptions of a monorepo being ultimately beneficial and the decision to migrate being finalised in the above discussion, proceeds to go into the implementation details.
+
+### Exclusive benefits of monorepo
+
+This only includes things that cannot be accomplished without the use of a monorepo.
+
+1. Single place to go for issues, PRs and all activity. Currently tickets are scattered across several repos, and any tickets that could benefit more than a single layer must be opened in each of the different repos.
```

Comment:
Not something exclusive of a monorepo, you can already [view all the issues of Openverse repositories aggregated in a single page](https://github.com/issues?q=is%3Aopen+is%3Aissue+repo%3AWordPress%2Fopenverse-catalog+repo%3AWordPress%2Fopenverse+repo%3AWordPress%2Fopenverse-api+repo%3AWordPress%2Fopenverse-frontend+repo%3AWordPress%2Fopenverse-infrastructure), including the private infrastructure repo.

Reply:
- What kind of bulk actions do you tend to do or would like to be able to do? I'm trying to imagine the use for mixed issues/PRs.
- I think the issue with the `gh` CLI can probably be solved with `just` commands on the `openverse` repo, which would require some configurations of course, but do you need to see often _all_ the open issues? For development one only typically need to focus on a well-defined subset.
- It's part of the GitHub UI in the sense that the link is always on the GH header, you only need to add the filters for the repos. I don't think new contributors need to look at all the issues to start either, approaching one individual level of the stack is more manageable.

Reply:
@krysal I don't disagree with your comments. There are potential solutions and workarounds for many of the things that a monorepo offers. I don't think the benefits of a monorepo need to be things that _only a monorepo can do_, but things that a _monorepo might do a bit better_ also need to be considered. 

Regarding contributors looking at lists of issues, I don't think they need to see _all_ issues, either. But consider the ways users might look at the issues page. Speaking for myself, when I hear about a new project I'd like to contribute to I tend to:

- Click on the 'issues' tab in the GitHub UI to look at the top 5-10 issues and see what's actively being discussed, worked on, and engaged with in a project
- Sort by "most commented" to see what the biggest or most popular discussions are

I think it's fundamentally simpler for a contributor who wants to look at Openverse issues to go to one repo and click on the "issues" tab in the UI, rather than type filters or click on a link which was found externally.

Reply:
> There are potential solutions and workarounds for many of the things that a monorepo offers

For me it isn't just this, but we should also be clear that there are trade-offs to using a monorepo, both long term and temporary. To me it's important that we agree that any benefits we think we'll get from it are worth the trade offs. That's why the conversation cannot just be "what are the benefits", it must be "what are the benefits and why are they worth making the change rather than working with what we have now". Monorepos can make a lot of meta infrastructural things more complex. The ci_cd workflow in the API repository is already complicated. A lot of actions will now run that don't need to, like Playwright for changes to the API. Playwright takes a long time to run. Is that a trade off we want? If we don't, we'll need to make a files filter, but that can only be applied to an entire workflow, so now we have to have separate workflows for different aspects of the projects, which is something we intentionally worked away from at one point. Is that something we are willing to accept?

A change like this is not just pure, uncomplicated benefits. It's not that we can do everything we're doing now but easier. There will be trade offs. Even just a bigger repository to check out can be an issue, more history to dig through, more complicated releases, etc.